File: src/Other/ObservableObject.vb
Imports System.ComponentModel
Imports System.Runtime.CompilerServices

Namespace OpenSilver.Samples.Showcase

    Public MustInherit Class ObservableObject
        Implements INotifyPropertyChanged

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Sub OnPropertyChanged(<CallerMemberName> Optional propertyName As String = Nothing)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub

        Protected Function SetProperty(Of T)(ByRef field As T, newValue As T, <CallerMemberName> Optional propertyName As String = Nothing) As Boolean
            If EqualityComparer(Of T).Default.Equals(field, newValue) Then
                Return False
            End If

            field = newValue
            OnPropertyChanged(propertyName)
            Return True
        End Function
    End Class

End Namespace

File: src/Other/ObservableObject.fs
namespace OpenSilver.Samples.Showcase

open System.Collections.Generic
open System.ComponentModel
open System.Runtime.CompilerServices

[<AbstractClass>]
type ObservableObject() =

    let propertyChanged = new Event<PropertyChangedEventHandler, PropertyChangedEventArgs>()

    interface INotifyPropertyChanged with
        [<CLIEvent>]
        member _.PropertyChanged = propertyChanged.Publish

    member this.OnPropertyChanged([<CallerMemberName>] ?propertyName: string) =
        let name = defaultArg propertyName null
        propertyChanged.Trigger(this, PropertyChangedEventArgs(name))

    member this.SetProperty<'T>(field: byref<'T>, newValue: 'T, [<CallerMemberName>] ?propertyName: string) : bool =
        if EqualityComparer<'T>.Default.Equals(field, newValue) then
            false
        else
            field <- newValue
            this.OnPropertyChanged(?propertyName = propertyName)
            true
